File: lintcode/remove_nth_node_from_end_of_list/py/remove_nth_node_from_end_of_list.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/remove-nth-node-from-end-of-list
@Language: Python
@Datetime: 17-02-16 00:00
'''

"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @param n: An integer.
    @return: The head of linked list.
    """
    def removeNthFromEnd(self, head, n):
        if head == None:
            return None
        
        dummy = ListNode(0, head)
        slow  = dummy
        fast  = dummy
        
        while n > 0:
            fast = fast.next
            n   -= 1
            
        while fast.next != None:
            slow = slow.next
            fast = fast.next
            
        slow.next = slow.next.next
        
        return dummy.next

File: lintcode/easy/remove_nth_node_from_end_of_list/cpp/remove_nth_node_from_end_of_list.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/remove-nth-node-from-end-of-list
@Language: C++
@Datetime: 17-02-16 00:05
*/

/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @param n: An integer.
     * @return: The head of linked list.
     */
    ListNode *removeNthFromEnd(ListNode *head, int n) 
    {
        auto dummy = ListNode{ 0 };
        dummy.next = head;
        
        auto slow  = &dummy;
        auto fast  = &dummy;
        
        while (n-- > 0) {
            fast = fast->next;
        }
        
        while (fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next;
        }
        
        auto next = slow->next->next;
        delete slow->next; slow->next = next;
        
        return dummy.next;
    }
};
